### PR TITLE
[7.x] 3.5x faster micro performance enhancement for data_get()

### PR DESCRIPTION
**Pull request by kevindees**
Replace `while:array_shift` with `foreach:unset`. This operation is 3.5x faster. It is arguably more readable as well.

In my isolation tests with 1000 iterations:

- While loop `1.969ms`
- Foreach loop `0.527ms`

### CODE REVIEW

**Comment by browner12** [comment]
how are you accounting for possible `null`s in the keys?

**Comment by GrahamCampbell** [comment]
> how are you accounting for possible nulls in the keys?

That never happens. Keys are either strings or integers. If you do an array access with another type of scalar, PHP will silently convert it (if it's numeric, to an integer then to a string, otherwise, straight to a string), and otherwise, it will raise a warning (or on PHP 8 I think it raises an actual Error).

**Comment by browner12** [comment]
correct, but `$keys` is a misleading variable name here.  We're not actually pulling "keys" from the array, we're pulling "values" from the `$key` array.

**Comment by kevindees** [comment]
> how are you accounting for possible `null`s in the keys?

I was not sure of the purpose of `is_null()`. I'm assuming it accounts for calls like this:

```php
    $data = [
        'product-one' => ['name' => ['sku' => 123], 'sku' => 987],
        'product-two' => ['name' => ['sku' => 123], 'sku' => 987],
    ];

    dd(data_get($data, ['*', null, 'sku']));
```

I have added a check for this but I'd assume this issue is an edge case. Maybe it should throw an exception instead. I'd imagine anyone sending `null`s is not getting the result they desire in the first place unless they are building into the edge case for some add reason.

**Comment by browner12** [comment]
The purpose of it originally was to figure out when the `$key` variable was empty.

According to the `array_shift` docs:

> Returns the shifted value, or NULL if array is empty or is not an array.

This way, the `while` loop would end once `array_shift` pulled all the items out of `$key`.

As I'm looking back at the old code, having a `null` value in your `$key` array would actually cause a premature termination of the `while` loop.  Based on that, I would say we don't actually have to check for it in the new code, so you can probably revert your latest commit.

```php
while(array_shift([null, 1, null])) {
    echo 'test';
}
```

In this example, "test" will never be echoed.

**Comment by kevindees** [comment]
@browner12 

Based on my testing adding the `null` check adds some sanity to the return value (in the edge-case) and the `null` check has little performance impact allowing the performance to remain intact. 

However, on the other hand, the `null` check can hide more serious bugs because it produces a more reasonable result. Maybe it would be removed in v8 but remain in v7 to keep backward compatibility.

**Comment by taylorotwell** [comment]
So after this discussion I'm confused. Is there *ANY* behavior change here at all?

**Comment by browner12** [comment]
I believe the latest commit that added the NULL check makes this code functionally equivalent to the old code. I ran a couple scenarios, but would love for a second set of eyes on it.

File: src/Illuminate/Support/helpers.php [hunk 1]
```diff
@@ -142,7 +142,13 @@ function data_get($target, $key, $default = null)
 
         $key = is_array($key) ? $key : explode('.', $key);
 
-        while (! is_null($segment = array_shift($key))) {
+        foreach ($key as $i => $segment) {
+            unset($key[$i]);
+
+            if ($segment === null) {
```
Comment (browner12) [inline src/Illuminate/Support/helpers.php:148]: there was a PR awhile ago that switched `$var === null` to `is_null($var)`.

https://github.com/laravel/framework/pull/15655/files

Might want to switch this line to maintain consistency.

Comment (kevindees) [reply src/Illuminate/Support/helpers.php:148]: This PR is in feature parody with the current 7.x state.

**Review by kevindees — COMMENTED** [review]
> So after this discussion I'm confused. Is there _ANY_ behavior change here at all?

There is no behavior change. It is in parity.

File: src/Illuminate/Support/helpers.php (hunk 1 above)
Comment (browner12) [reply src/Illuminate/Support/helpers.php:148]: I agree the behavior is the same, but to maintain style consistency, we should switch `$segment === null` to `is_null($segment)`.

Also, FYI. *Parity*, not *parody*.

Comment (kevindees) [reply src/Illuminate/Support/helpers.php]: Done

**Comment by taylorotwell** [comment]
Thank you
